CDCgov/prime-reportstream
# 2751 Integrate database lookup tables

This PR integrates the database lookup tables to be handled just like file-based tables.  The result is that tables can be either file-based or databased backed.  File-based take precedence over database.  To reload the database lookup tables, the use of the metadata singleton is used to trigger a check on that last time the database was checked, and when that time expires then it looks to the database to check for any updates.

Test Steps:
We are going to be using the LIVD table to test that the database tables can be used instead of file-based tables as well as loading updates.
1. Start a terminal and go to the `prime-router` folder.
2. Run your local RS instance.
3. Send the `./src/testIntegration/resources/datatests/CSV_to_HL7/sample-single-pdi-20210608-0002.csv` test file to the reports endpoint using the `simple_report` client.
4. Wait until the file is fully processed and verify that the pima-az-covid file sent to the `build/sftp` folder has an Ordered_test_name of `SARS-CoV-2 (COVID-19) Ag [Presence] in Respiratory specimen by Rapid immunoassay`.  This is the baseline we will compare to for the following tests.
5. Create a new lookup table by running the following command: 
`./prime lookuptables create -n LIVD-SARS-CoV-2-2021-09-29 -i ./metadata/tables/LIVD-SARS-CoV-2-2021-09-29.csv`
and answering Y at the prompt.
6. Shutdown your RS instance.
7. Move the LIVD table file we used to the prime-router folder by running 
`mv ./metadata/tables/LIVD-SARS-CoV-2-2021-09-29.csv ./`
This will remove that table's file from the system and allow RS to load the table from the database.
8. Run `gradle clean package` and then start your local RS instance.
9. Send the `./src/testIntegration/resources/datatests/CSV_to_HL7/sample-single-pdi-20210608-0002.csv` test file again to the reports endpoint using the `simple_report` client.
10. Wait until the file is fully processed and verify that the latest pima-az-covid file sent to the `build/sftp` folder has an Ordered_test_name of `SARS-CoV-2 (COVID-19) Ag [Presence] in Respiratory specimen by Rapid immunoassay`. 
11. Edit the file `./LIVD-SARS-CoV-2-2021-09-29.csv`, find the record for `BinaxNOW COVID-19 Ag Card` and change the column `Test Ordered LOINC Long Name` to a different text like `MY TEST SARS-CoV-2 (COVID-19) Ag [Presence] in Respiratory specimen by Rapid immunoassay`. 
12. Create a new lookup table by running the following command: `./prime lookuptables create -n LIVD-SARS-CoV-2-2021-09-29 -i ./LIVD-SARS-CoV-2-2021-09-29.csv` then answer Y at the prompt.
13. Activate the new version by running `./prime lookuptables activate -n LIVD-SARS-CoV-2-2021-09-29 -v 2` (replace the number 2 in `-v 2` with the proper new version if not 2). 
14. Wait 40 seconds to allow for the poll timer to expire, so the new table is loaded on the next run.
15. Send the `./src/testIntegration/resources/datatests/CSV_to_HL7/sample-single-pdi-20210608-0002.csv` test file to the reports endpoint using the `simple_report` client.
16. Wait until the file is fully processed and verify that the latest pima-az-covid file sent to the `build/sftp` folder has an Ordered_test_name of `MY TEST SARS-CoV-2 (COVID-19) Ag [Presence] in Respiratory specimen by Rapid immunoassay`.
17. Revert your repo back to normal by running the following commands:
`rm ./LIVD-SARS-CoV-2-2021-09-29.csv`
`git checkout origin/master ./metadata/tables/LIVD-SARS-CoV-2-2021-09-29.csv`

## Changes
- Integrate database lookup tables - this is done loading the data as Lookup Table objects, which are then handled the same way as the file-based tables.
- Moved the table classes into a metadata folder, aiming to move more in the future to organize the code a little bit more
- Made Metadata a singleton - previously Metadata was being called everywhere by its constructor that loaded everything every time.  Tried to make as much as possible private, so this does not happen again.
- Check for table updates within the Metadata getinstance method.  We are unable to have an outside trigger reload the tables as we have multiple threads and instances running.  Every function that needs metadata will have to invoke the getInstance() function which will trigger a table check at an interval.
 - - Blank out tables that are no longer in the DB or active as removing from memory may break things
 - - Add any newly activated tables to open to a future with some dynamic changes to other portions
- Some other minor cleanups

## Checklist

### Testing

### Process

## Fixes
- #2751

## To Be Done
- #2885 -  need to load test data into the DB for local use

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
